### PR TITLE
[Refactor] Reduce ops/ file count: merge avg_pool, drop _fwd/_bwd suffixes, add attention/ subpackage

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

- Merge `avg_pool1d.py`, `avg_pool2d.py`, `avg_pool3d.py` into `pool.py`
- Merge `conv1d.py`, `conv2d.py`, `conv3d.py` into `convolution.py`
- Merge `engram_fwd.py` + `engram_bwd.py` into `engram.py`
- Rename 6 standalone `*_fwd.py` files to drop the `_fwd` suffix (`da_cumsum`, `ssd_chunk_scan`, `ssd_chunk_state`, `ssd_state_passing`, `gqa_sliding_window`, `gqa_sliding_window_varlen`)
- Move 13 attention op files (gqa, mha, deepseek_mla/dsa/nsa, mhc) into new `ops/attention/` subpackage
- Update manifest `source.op` paths for moved files
- Update all imports in `__init__.py`, tests, and benchmarks to match new paths

No functional changes. All public API symbols remain unchanged. Net reduction of ~19 files from `tileops/ops/` root.

## Test plan

- [x] pre-commit passed
- [x] `python -c "from tileops.ops import GroupedQueryAttentionFwdOp, MultiHeadAttentionFwdOp, AvgPool1dOp, Conv2dOp, DaCumsumFwdOp, SsdChunkScanFwdOp"` — all imports resolve
- [ ] pytest full suite (CI)

## Regression

Pure file-move and import-path refactor. No code logic touched. Any import breakage will surface immediately as `ImportError` in CI.

## Follow-up

No follow-up issues or suggestions.

### CODE REVIEW

File: tileops/ops/__init__.py [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
     NSATopkVarlenOp,
 )
+from .avg_pool import AvgPool1dOp, AvgPool2dOp, AvgPool3dOp
+from .conv1d import Conv1dBiasFwdOp, Conv1dFwdOp
+from .conv2d import Conv2dOp
+from .conv3d import Conv3dOp
+from .da_cumsum import DaCumsumFwdOp
 from .deltanet_chunkwise import DeltaNetBwdOp, DeltaNetFwdOp, DeltaNetOp
 from .deltanet_recurrence import DeltaNetDecodeOp
 from .dropout import DropoutOp
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/__init__.py:25]: Not a regression. `EngramGateConvFwdOp` and `EngramGateConvBwdOp` were never exported from `tileops.ops.__init__` — they were always imported directly from the submodule (e.g. `from tileops.ops.engram_fwd import EngramGateConvFwdOp`). This PR preserves that: they are now at `tileops.ops.engram`. No public API change.

File: tileops/ops/pool.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
+from typing import Dict, Optional, Tuple
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/pool.py:1]: No change needed. This code was moved verbatim from `avg_pool1d.py` which already used `int | tuple[int]` syntax (PEP 604). This is a move-only refactor — style normalization is out of scope.

File: tileops/ops/pool.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    """
+
+    def __init__(
+        self,
+        n: int,
+        c_in: int,
+        l_in: int,
+        kernel_size: int | tuple[int],
+        stride: Optional[int | tuple[int]] = None,
+        padding: int | tuple[int] = 0,
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/pool.py:33]: Same as above — preserving existing code from `avg_pool1d.py` verbatim. Type hint style normalization is out of scope for this move-only refactor.

File: tileops/ops/pool.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+    def __init__(
+        self,
+        n: int,
+        c_in: int,
+        h_in: int,
+        w_in: int,
+        kernel_size: int | Tuple[int, int],
+        stride: Optional[int | Tuple[int, int]] = None,
+        padding: int | Tuple[int, int] = 0,
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/pool.py:107]: Existing code from `avg_pool2d.py`, moved without modification. Style consistency across the file is a valid concern but out of scope for this structural refactor.

File: tileops/ops/engram.py [hunk 5]
```diff
@@ -18,6 +18,114 @@ def _align_up(n: int, alignment: int) -> int:
...
+            rms_w_v = F.pad(rms_w_v, (0, self.d_padded - self.d))
+            conv_w = F.pad(conv_w, (0, self.d_padded - self.d))
+
+        results = self.kernel(H, k, v, rms_w_h, rms_w_v, conv_w)
+        # results: [Y, vhat, alpha, rrms_h, rrms_k, rrms_v]
+
+        # Trim padding on d-dim tensors
+        if self.d_padded != self.d:
+            results[0] = results[0][:, :, :self.d]   # Y
+            results[1] = results[1][:, :, :self.d]   # vhat
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/engram.py:124]: Existing code from `engram_fwd.py`, moved without modification. The kernel returns a list — this has been working in production. Adding a defensive `list()` wrapper is a behavior change outside the scope of this move-only refactor.

File: tileops/ops/pool.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
+from typing import Dict, Optional, Tuple
+
+import torch
+
+from tileops.kernels.kernel import Kernel
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/pool.py:5]: Adopted. Updated PR description to reflect the actual file names (`pool.py`, `convolution.py`) instead of the stale `avg_pool.py` reference.

File: tests/ops/test_da_cumsum_fwd.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import torch
 
 from tests.test_base import TestBase
-from tileops.ops.da_cumsum_fwd import DaCumsumFwdOp
+from tileops.ops.da_cumsum import DaCumsumFwdOp
```
Comment (Ibuki-wind) [inline tests/ops/test_da_cumsum_fwd.py:4]: This implementation refactor updates a Test-stage file in the same PR, which violates the trust model boundary for `tests/` -> split the `tests/` import-path updates into a follow-up test-owned PR.

File: benchmarks/ops/bench_da_cumsum_fwd.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 import torch
 
 from benchmarks.benchmark import BenchmarkBase, BenchmarkReport
-from tileops.ops.da_cumsum_fwd import DaCumsumFwdOp
+from tileops.ops.da_cumsum import DaCumsumFwdOp
```
Comment (Ibuki-wind) [inline benchmarks/ops/bench_da_cumsum_fwd.py:7]: This implementation refactor updates a Benchmark-stage file in the same PR, which violates the trust model boundary for `benchmarks/` -> move the benchmark import-path churn into a separate benchmark-owned PR after the ops layout change lands.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Trust-boundary verdict: blocking.
This refactor mixes Implementation (`tileops/ops/`) with Test (`tests/`) and Benchmark (`benchmarks/`) stage edits in one PR, which violates the project trust model.
The op/package reorganization itself looks mechanically consistent, but it needs to land as an implementation-only PR.
Move the import-path updates in `tests/` and `benchmarks/` into follow-up stage-owned PRs after the implementation layout is settled.
Main risk: allowing cross-stage refactors here weakens the review boundary the repo relies on for interface and correctness discipline.

File: tests/ops/test_da_cumsum_fwd.py (hunk 7 above)
Comment (lcy-seso) [reply tests/ops/test_da_cumsum_fwd.py:4]: This is checked by human developer.

Comment (lcy-seso) [reply tests/ops/test_da_cumsum_fwd.py:4]: The trust model boundary prevents code PRs from weakening test assertions. This PR makes zero logic changes — only import paths are updated to match moved modules. Splitting these into a follow-up would leave tests with broken imports on main between merges, which is worse than co-shipping the mechanical fix.

File: benchmarks/ops/bench_da_cumsum_fwd.py (hunk 8 above)
Comment (lcy-seso) [reply benchmarks/ops/bench_da_cumsum_fwd.py:7]: Same reasoning as the test imports above. These are mechanical import-path updates required by the file moves. No benchmark logic, thresholds, or workloads are modified. Splitting would break benchmarks on main between PRs.

Comment (lcy-seso) [reply benchmarks/ops/bench_da_cumsum_fwd.py:7]: This is checked by human developer

File: tests/ops/test_da_cumsum_fwd.py (hunk 7 above)
Comment (Ibuki-wind) [inline tests/ops/test_da_cumsum_fwd.py:4]: This implementation refactor still edits a Test-stage file, which violates the trust-model boundary for `tests/` -> move the test import-path update into a separate test-owned PR after the implementation layout lands.

File: benchmarks/ops/bench_da_cumsum_fwd.py (hunk 8 above)
Comment (Ibuki-wind) [inline benchmarks/ops/bench_da_cumsum_fwd.py:7]: This implementation refactor still edits a Benchmark-stage file, which violates the trust-model boundary for `benchmarks/` -> move the benchmark import-path update into a separate benchmark-owned PR after the implementation layout lands.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Trust-boundary verdict: blocking.
This refactor still mixes Implementation (`tileops/ops/`) with Test (`tests/`) and Benchmark (`benchmarks/`) stage edits in one PR, which violates the project trust model.
The package move itself looks mechanically consistent, but it needs to land as an implementation-only PR.
Split the `tests/` and `benchmarks/` import-path updates into follow-up stage-owned PRs after the ops layout is stable.
Main risk: accepting cross-stage refactors here weakens the review boundary this repo uses to keep interface, correctness, and performance concerns independent.

File: tileops/ops_manifest.yaml [hunk 9]
```diff
@@ -553,7 +553,7 @@ ops:
       kernel: tileops/kernels/flash_attn/fwd.py
       kernel_map:
         mha_fwd_kernel: MhaFwdKernel
-      op: tileops/ops/mha.py
+      op: tileops/ops/attention/mha.py
```
Comment (Ibuki-wind) [inline tileops/ops_manifest.yaml:556]: This implementation refactor also edits the Manifest stage, but `tileops/ops_manifest.yaml` is source-of-truth metadata that requires a separate human-reviewed PR -> move the manifest path rewrites into their own manifest-only PR.

File: tests/ops/test_da_cumsum_fwd.py (hunk 7 above)
Comment (Ibuki-wind) [inline tests/ops/test_da_cumsum_fwd.py:4]: This implementation refactor still edits a Test-stage file, which violates the trust-model boundary for `tests/` -> move the test import-path update into a separate test-owned PR after the implementation layout lands.

File: benchmarks/ops/bench_da_cumsum_fwd.py (hunk 8 above)
Comment (Ibuki-wind) [inline benchmarks/ops/bench_da_cumsum_fwd.py:7]: This implementation refactor still edits a Benchmark-stage file, which violates the trust-model boundary for `benchmarks/` -> move the benchmark import-path update into a separate benchmark-owned PR after the implementation layout lands.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Trust-boundary verdict: blocking.
This refactor now spans Manifest (`tileops/ops_manifest.yaml`), Implementation (`tileops/ops/`), Test (`tests/`), and Benchmark (`benchmarks/`) in one PR, which violates the project trust model.
The path rewrites themselves look mechanically consistent, but they need to be split by stage ownership.
Land the implementation-only package move first, then stage-owned follow-ups for test imports and benchmark imports, and keep manifest path changes in a separate human-reviewed PR.
Main risk: merging this as-is weakens the repo’s review boundary between source-of-truth spec, implementation, correctness, and performance layers.

**Review by Ibuki-wind — APPROVED** [review]
Treating the confirmed trust-boundary break as intentional, this refactor looks mechanically consistent. I did not find an additional design, interface, or correctness issue in the path moves, `__init__` rewires, or manifest path updates.
